proposals/NNNN-package-manager-product-definitions.md

```diff
@@ -0,0 +1,293 @@
+
+Any target may be included in multiple products, though not all kinds of targets apply to all products; for example, a test target is not able to be included in a library product.
+
+The products represent the publicly vended package outputs on which any client package can depend, and also define the characteristics of the produced artifacts.
+
+As mentioned earlier, Swift Package Manager already infers a set of implicit products based on the set of targets in the package.  One open question is whether inference rules will be supported and documented as part of this feature, in order to allow existing packages to continue to work unmodified and to avoid the need for explicit product definitions for simple packages.  While the inference rules will need to be supported for existing packages, it is still an open question whether product will be inferred for packages using the Swift Pacakge Manager 4.0 version of the API.
+
+The inference rules that apply to legacy packages are detailed in the next section of this document.
+
+Even if Swift Pacakge Manager 4.0 does infer products for non-legacy packages, products will only be inferred if there are no explicit product definitions in the package.  The reason is that the interaction of a mixture of implicit and explicit products is likely to lead to significant complexity and to be a source of both confusion and bugs.
+
+An example of a package containing explicit product definitions:
+
+```
+let package = Package(
+    name: "Hello",
+    targets: [
+        Target(name: "Foo"),
+        Target(name: "Bar"),
+        Target(name: "Baz", dependencies: ["Foo", "Bar"]),
+        Target(name: "Exe", dependencies: ["Foo"])
+    ],
+    products: [
+        Library(name: "Lib1", type: .static, targets: ["Bar"]),
+        Library(name: "Lib2", type: .dynamic, targets: ["Baz"]),
+        Executable(name: "Exe", targets: ["Exe"]),
+    ]
+)
+```
+
+The initial types of products that can be defined are executables and libraries.  Libraries can be declared as static or dynamic, but when possible, the specific type of library should be omitted -- in this case, the build system will choose the most appropriate type to build based on the context in which the product will be used (depedning on the type of client, the platform, etc).
+
+Note that tests are not considered to be products, and do not need to be explicitly defined.
+
+A product definition lists the root targets to include in the product; the interfaces of those targets will be available to any clients (for product types where that makes sense).  Any dependencies of those targets will also be included in the product, but won't be made visible to clients.  The Swift compiler does not currently provide this granularity of visibility control, but it constitutes a declaration of intent that can be used by IDEs and other tools.  We also hope that the compiler will one day support this level of visibility control.
```

Comment:
We can actually enforce this if, for a particular target, we only add -I (import path) to the swiftmodule files of the root targets of the product